### DISCUSSION TITLE
Virtual matter device

### DISCUSSION THREAD

**Discussion by MartinsMihaeljans** [FezVrasta/node-red-matter#9] [Q&A] [answered]
Hello, this Matter integration with Node-RED seems really interesting to me. However, I'm not able to deploy the example flow as I have no physical matter devices in my network. So, my question is - **Is there a way to emulate a Matter device on PC to work with this node-RED project?**

**Comment by FezVrasta** [comment, marked as answer]
Hi, the whole purpose of this project is to emulate Matter devices on Node-RED. You do need a Matter compatible controller though to pair the devices

**Reply by MartinsMihaeljans** [reply]
First of all, thank you for such quick response, @FezVrasta. I now have another question - is there a way I can emulate the controller too? Perhaps on another PC or Android device? Or it must be a physical device? And if so what device would you suggest?

**Reply by FezVrasta** [reply]
I'm not sure what you are after? Could you provide an example?

**Reply by MartinsMihaeljans** [reply]
What I was thinking of was attaching some http-request nodes to Matter switches and export the JavaScript in to a  stand alone software. Sort of creating a digital IoT object which would exist in Matter network but not in the physical area. I wanted to test if this is possible but without making any big purchases :)

**Reply by FezVrasta** [reply]
And why do you need a controller? Controllers are Apple Homekit, Google Home, Alexa, etc

**Reply by MartinsMihaeljans** [reply]
From your first response I assumed that I need a controller to even deploy the switches. Is that right?

**Reply by FezVrasta** [reply]
No you can deploy them just with this node, but you can't "use them" (from a smartphone for example) without first pairing them to a controller

**Reply by MartinsMihaeljans** [reply]
Great, thank you for sparing time. @FezVrasta What do you think about this project - https://github.com/home-assistant-libs/python-matter-server would this work with node-RED?

**Reply by FezVrasta** [reply]
python and node are different technologies so the integration wouldn't be straight forward

**Reply by Apollon77** [reply]
matter.js (the base library of the nodes) also has a "controller shell app" to use as controller. not really that many features as of now, but increasing

**Reply by MartinsMihaeljans** [reply]
@Apollon77  Could you tell me how to run the controller shell app? Or at least guide me to some tutorial?

**Reply by Apollon77** [reply]
https://github.com/project-chip/matter.js/tree/main/packages/matter-node-shell.js... but you need the nightly build for that from npm ... it is not included yet in the officvial npm version. so install with @dev

**Reply by MartinsMihaeljans** [reply]
@Apollon77 @FezVrasta Thank you very much. This helps a lot. I'll try to put this to work and get beck with the results :)

**Comment by MartinsMihaeljans** [comment]
Hello, @FezVrasta and @Apollon77. Sorry for the long delay. I failed to install the controller shell app. My setup is as follows:
VirtualBox VM running Ubuntu 22 desktop OS;
Git repo branch selected nightly-release;
When I try **npm install** command this is the outcome:
 
<img width="743" alt="image" src="https://github.com/FezVrasta/node-red-matter/assets/128288466/ab774089-7d05-4db1-9347-7a9014c93ef8">

And here is a full log file.
[log.txt](https://github.com/FezVrasta/node-red-matter/files/13378097/log.txt)

**What could be the cause of this? Does my virtual machine or physical pc Bluetooth setup needs changes?**

**Reply by Apollon77** [reply]
the cause is that there are buld issues with the ble package. could it be that python 3.12 is installed? If yes two options:

Downgrade python OR upgrade npm (10.2.2+)

The cause is a change in lython 3.12 which triggers a issue in node-gyp which means builds of native components fail.

**Reply by MartinsMihaeljans** [reply]
Okay, now I got a little further:
<img width="733" alt="image" src="https://github.com/FezVrasta/node-red-matter/assets/128288466/4f5ba0bc-398e-45c2-8aeb-ac515e727aec">
npm and python versions are as follows:
<img width="374" alt="image" src="https://github.com/FezVrasta/node-red-matter/assets/128288466/7bc077e9-5bc4-4eed-908f-d04ba4a7f5e0">

I tried to do npm install jsdom but it was no help.
@Apollon77 What could be a cause of this?

**Reply by Apollon77** [reply]
Ok then maybe it is a different reason ... try to install the packet with " --foreground-scripts" as parameters. whats the log then ? (and pleas epost text and no screenshots)

**Reply by MartinsMihaeljans** [reply]
Hi, @Apollon77 and @FezVrasta I got it running thanks to parameter --foreground-scripts. Now I have another question. I understand that I have run an controller instance but do I need to configure some addressing for it? Also, how do I set the use of my shell controller in node-red matter example, as in debug it says that it is not set?
<img width="918" alt="image" src="https://github.com/FezVrasta/node-red-matter/assets/128288466/c1469b8c-bb84-4655-ae60-d4e26fb2559b">

**Reply by Apollon77** [reply]
> got it running thanks to parameter --foreground-scripts

Interesting ... this normally just makes sure that we see logging :-)) but ok if it worked fine.

To your question: You want to have a device in node-red and control this using the shell app? Then in fact you need to configzre the devcie and start that in node-red.

Then in the shell app you use the "commission" commands to pair/commission the devcie to the shell-controller ... check help output there for commands and options.
